go/types/sequence_cursor.go

```diff
@@ -32,16 +49,67 @@ func (cur *sequenceCursor) getItem(idx int) sequenceItem {
 	return cur.seq.getItem(idx)
 }
 
+// sync loads the sequence that the cursor index points to.
+// It's called whenever the cursor advances/retreats to a different chunk.
 func (cur *sequenceCursor) sync() {
 	d.PanicIfFalse(cur.parent != nil)
 	cur.seq = cur.parent.getChildSequence()
+	if cur.readAheadMap != nil {
+		// if this cursor uses read-ahead, reset the cache for the next chunk
+		cur.readAheadMap = make(map[int]chan sequence)
+		cur.readAheadNext = 0
+	}
+}
+
+var readAheadParallelism = runtime.NumCPU() * 8
+
+func minInt(x, y int) int {
+	return int(math.Min(float64(x), float64(y)))
 }
 
+func maxInt(x, y int) int {
+	return int(math.Max(float64(x), float64(y)))
+}
+
+// readAhead (called when read-ahead is enabled) primes the next entries in the
+// read-ahead cache. It ensures that go routines have been allocated for reading
+// the next n entries in the current sequence. N is either readAheadParallelism
+// or the number of entries left in the sequence if smaller.
+func (cur *sequenceCursor) readAhead() {
+	d.PanicIfTrue(cur.readAheadMap == nil)
+	// the next position to be primed
+	cur.readAheadNext = maxInt(cur.idx, cur.readAheadNext)
+	// the last position to be primed
+	last := minInt(cur.idx+readAheadParallelism, cur.seq.seqLen()) - 1
+	// prime all positions between next and last
+	for i := cur.readAheadNext; i <= last; i += 1 {
+		futureSeq := make(chan sequence)
```

Comment:
I actually think we want a buffer size of 1 here, otherwise the goroutines won't terminate until the main thread reads off the channel, which might not happen if the caller never finishes iteration.

If it's 1 then the goroutine will immediately exit, and the channel's buffer will just be GC'd with the sequence.